MGMT-6410: Adding documentation for OCM integration. (#1795)
openshift/assisted-service: docs/ocm-integration.md

# OCM Integration

> Assisted-service uses the [OCM(OpenShift Cluster Manager)](https://openshift.github.io/openshift-origin-design/designs/ocm/) client in order to make AI clusters shown as part of a users' clusters in the OCM UI.

Reviewer: ```suggestion
OpenShift Cluster Manager (OCM) is a managed service where users can install, operate and upgrade Red Hat OpenShift 4 clusters.  This document describes the integration was done such that Assisted Installed (AI) clusters are shown as part of a user's clusters in the OCM UI.
To achieve this, the assisted-installer users the [OCM client](https://github.com/openshift-online/ocm-sdk-go) to make several calls during a cluster's installation.
```

## AMS

AMS is the micro-service in OCM which holds the users' clusters list, which is indeed a list of `Subscriptions` AMS objects, and handles the authZ for assisted-service API calls.

> Sometimes when people refer to OCM, they actually mean AMS inside OCM.

Reviewer: This sentence can be deleted.

Author: Done.

## Cluster lifecycle

### Cluster registration

On cluster registration, the service will create an AMS subscription for the cluster with some initial values:
```
status: "Reserved"
cluster_id: The cluster id registered in AI DB.
display_name: The cluster name in AI DB.
product_id: "OCP"
product_category: "AssistedInstall"
```

### Cluster is updated with a new cluster-name

In this case the service will patch the subscription with the new cluster name.
```
display_name: New cluster name
```

### Cluster installation

During the cluster installation process we will have multiples point of integration with AMS.

Once an `openshift_cluster_id` is generated during the `preparing-for-installation` state, the service will patch the subscription with:
```
external_cluster_id: Openshift cluster ID
```

Later on, when console operator is installed during the `finalizing` state, the service will patch the subscription with:
```
console_url: Cluster's console URL
```

Finally, when the cluster is fully installed (or degraded), the service will patch the subscription with:
```
status: "Active"
```

### Cluster is sending metrics to Openshift Telemeter

> Once metrics reach the Telemeter server, Telemeter will notify AMS and AMS will search for a subscription with a matching `external_cluster_id` (which is included in the metrics the cluster send). If it finds such a subscription it will add a `metrics` field to the subscription, otherwise, it will create a new subscription for that "unsubscribed" cluster but in this case, all the data patched by AI will be missing - this should be a bug indication. 

Reviewer: ```suggestion
Once metrics from the installed cluster reach the Telemeter server, Telemeter will notify AMS and AMS will search for a subscription with a matching `external_cluster_id` (which is included in the metrics that the cluster sends). If it finds such a subscription it will add a `metrics` field to the subscription, otherwise, it will create a new subscription for that "unsubscribed" cluster but in this case, all the data patched by AI will be missing - this should be a bug indication. 
```

### Cluster deletion

> On cluster deregistration, we will differ between 2 flows:

Reviewer: ```suggestion
On cluster deregistration, we have 2 flows:
```

#### Clusters with `Active` subscription

Those clusters are fully installed and are running on the clients' machines, therefore, when deleting those clusters from the service, whether is deleted by the user or by GC, the subscription is left untouched and the client will continue to see his cluster on the OCM UI.

AMS is refreshing subscriptions periodically and if a running cluster stop delivering metrics for some time it will change its subscription status to "Stale" and after a longer period to `Archived`

#### Clusters with `Reserved` subscription

> Those clusters where not installed for some reason and AMS won't monitor those subscriptions, therefore, it is AI responsibility to delete those subscriptions in AMS if the cluster is deleted from the service.

Reviewer: ```suggestion
Those clusters where not installed for some reason and AMS won't monitor those subscriptions, therefore, it is AI's responsibility to delete those subscriptions in AMS if the cluster is deleted from the service.
```

# OCM Integration

> Assisted-service uses the OCM client in order to make AI clusters shown as part of all your clusters in OCM UI.

Reviewer 2: ```suggestion
Assisted-service uses the [OCM(OpenShift Cluster Manager)](https://openshift.github.io/openshift-origin-design/designs/ocm/) client in order to make AI clusters shown as part of a users' clusters in the OCM UI.
```

Reviewer: ```suggestion
OpenShift Cluster Manager (OCM) is a managed service where users can install, operate and upgrade their Red Hat OpenShift 4 clusters.  This document describes the integration such that clusters being installed with Assisted Installer (AI) are displayed as part of a user's clusters in OCM.
The assisted-service uses the [OCM client](https://github.com/openshift-online/ocm-sdk-go) for this integration.
```

## AMS

> AMS is the micro-service in OCM which hold the users clusters list, which is indeed a list of `Subscriptions` AMS objects, and handles the authZ for assisted-service API calles.

Reviewer 2: ```suggestion
AMS is the micro-service in OCM which holds the users' clusters list, which is indeed a list of `Subscriptions` AMS objects, and handles the authZ for assisted-service API calls.
```

> Somethimes when people refers to OCM they actually mean AMS inside OCM.

Reviewer 2: ```suggestion
Sometimes when people refer to OCM, they actually mean AMS inside OCM.
```

Reviewer: This sentence can be deleted.

## Cluster lifecycle

### Cluster registration

> On cluster registration, the service will create an AMS subscription for the cluster which some initial values:

Reviewer 2: ```suggestion
On cluster registration, the service will create an AMS subscription for the cluster with some initial values:
```

Reviewer: ```suggestion
On cluster registration, the assisted-service will create an AMS subscription for the cluster with some initial values:
```

```
status: "Reserved"
cluster_id: The cluster id registered in AI DB.
display_name: The cluster name in AI DB.
product_id: "OCP"
product_category: "AssistedInstall"
```

> ### Cluster is updated with a new cluster-name
> 
> In this case the service will patch the subscription with the new cluster name.
> ```
> display_name: New cluster name
> ```

Reviewer: ```suggestion
### Cluster is renamed

In this case, the assisted-service will patch the subscription with the new cluster name.
```
display_name: New cluster name
```
```

> ### Cluster installation

Reviewer: What happens if installation fails?

Author: I will add this part to the doc.

Author: Done.

> During the cluster installation process we will have multiples point of integration with AMS.

Reviewer: ```suggestion
The assisted-service contacts AMS at several points during the cluster installation process.
```

Once an `openshift_cluster_id` is generated during the `preparing-for-installation` state, the service will patch the subscription with:
```
external_cluster_id: Openshift cluster ID
```

Later on, when console operator is installed during the `finalizing` state, the service will patch the subscription with:
```
console_url: Cluster's console URL
```

> Finally, when the cluster is fully installed (or degraded), the service will patch the subscription with:

Reviewer: ```suggestion
Finally, when the cluster is successfully installed, the service will patch the subscription with:
```

```
status: "Active"
```

> ### Clsuter is sending metrics to Openshift Telemeter

Reviewer 2: ```suggestion
### Cluster is sending metrics to Openshift Telemeter
```

> Once metrircs reaches Telemeter server, Telemeter will notify AMS and AMS will seach for a subscription with a matching `external_cluster_id` (it is included in the metrics the cluster send). If it finds such subscription it will add a `metrics` field to the subscription, otherwise, it will create a new subscription for that "unsubscribed" cluster but in this case all the data patched by AI will be missing - this should be a bug indication. 

Reviewer 2: ```suggestion
Once metrics reach the Telemeter server, Telemeter will notify AMS and AMS will search for a subscription with a matching `external_cluster_id` (which is included in the metrics the cluster send). If it finds such a subscription it will add a `metrics` field to the subscription, otherwise, it will create a new subscription for that "unsubscribed" cluster but in this case, all the data patched by AI will be missing - this should be a bug indication. 
```

### Cluster deletion

On cluster deregistration, we will differ between 2 flows:

#### Clusters with `Active` subscription

> Those clusters are fully installed and are running on the clients machines, therefore, when deleting those clusters from the service, whether is deleted by the user or by GC, the subscription is left untouched and the client will continue to see his cluster on the OCM UI.

Reviewer 2: ```suggestion
Those clusters are fully installed and are running on the clients' machines, therefore, when deleting those clusters from the service, whether is deleted by the user or by GC, the subscription is left untouched and the client will continue to see his cluster on the OCM UI.
```

AMS is refreshing subscriptions periodically and if a running cluster stop delivering metrics for some time it will change its subscription status to "Stale" and after a longer period to `Archived`

#### Clusters with `Reserved` subscription

Those clusters where not installed for some reason and AMS won't monitor those subscriptions, therefore, it is AI responsibility to delete those subscriptions in AMS if the cluster is deleted from the service.

## How to see the subscription in AMS using the OCM-cli

> You can sign-in to AMS using the ocm-cli in order to get information regarding your subscriptions.

Reviewer 2: ```suggestion
You can sign-in to AMS using the [ocm-cli](https://github.com/openshift-online/ocm-cli) in order to get information regarding your subscriptions.
```

> 
> First install ocm-cli:
> ```
> dnf copr enable ocm/tools
> dnf install ocm-cli
> ```
> 

Reviewer 2: ```suggestion
First, follow how to [install](https://github.com/openshift-online/ocm-cli#installation) ocm-cli.

```
